sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala

```diff
@@ -566,7 +566,7 @@ abstract class InterpretedHashFunction {
         hashUnsafeBytes(a, Platform.BYTE_ARRAY_OFFSET, a.length, seed)
       case s: UTF8String =>
         val st = dataType.asInstanceOf[StringType]
-        if (st.supportsBinaryEquality) {
+        if (st.supportsBinaryEquality && !st.usesTrimCollation) {
```

Comment:
on another note, let's make sure that we reconsider this in the follow-up:
https://issues.apache.org/jira/browse/SPARK-49911

please add `// TODO(SPARK-49911)` comments wherever appropriate, just so that we don't miss it

Reply:
I'm not really a fan of leaving comments like this in code. I will follow up with pr **today**, so won't miss them.